storage/docs/rest.md

```diff
@@ -1,69 +1,118 @@
+or more directories e.g:
+* ``` /files/libero/foo.txt ```
+* ``` /files/libero/baz/bar/foo.txt ```
+
+The key is unique within the namespace and provides a mapping between the storage entity and the provider entity (e.g. S3).
+
+## Storing a file
+
+To store a file with a namespace of ```libero``` and a key of ```directory/file.pdf```, the following request should be
+used
+
+```http
+PUT /files/libero/directory/file.pdf
 
-```
-PUT /files/namespace/directory/file.ext
 Content-Type: application/pdf
+Content-Length: 12345
 If-Match: <etag_value>|*
-Libero-file-tags: "key=value,key2=value2,filename=bar.pdf"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+```
 
-Response:
+The Server should respond with:
+
+```http
 Code: 201
-Link: "<http://url/foo/bar.pdf>;rel=signed"
+Link: "<http://storage-url/path/to/file>;rel=https://libero.pub/public"
 Last-Modified: 2019-08-19T14:40:04.123456
 ETag: <etag>
 ```
 
-## Get a file
+When uploading for the first time, the ```If-Match``` header should be set to '*'. If the ```If-Match``` fails,
+then a 412 (Precondition Failed) should be returned.
+
+If a there is no ```If-Match``` sent, then the storage service should send back a 428 (Precondition Required).
+
+## Retrieving a file
 
+To retrieve a file in the namespace ```libero``` and with key of ```directory/file.pdf```:
+
+```http
+GET /files/libero/directory/file.pdf
 ```
-GET /files/namespace/directory/file.ext
-#Prefer: original
 
 Response:
+```http
 Code: 200
 Content-Type: application/pdf
 Content-Length: 12345
-Libero-file-tags: "key=value,key2=value2,filename=bar.pdf"
-Link: "<http://sharedlink>;rel=????"
-Link: "<http://publiclink>;rel=????"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+Link: "<http://storage-url/path/to/file>;rel=https://libero.pub/public"
 Last-Modified: 2019-08-19T14:40:04.123456
 ETag: <etag>
+
 FILE DATA
 ```
 
-## Get file meta data
+The returned Link header value is the public url if the underlying storage is public.
+
+## Retrieving a file meta data
+
+### Public storage
 
+To retrieve the metadata of a file in the namespace ```libero``` and with key of ```directory/file.pdf```:
+
+```http
+HEAD /files/libero/directory/file.pdf
 ```
-HEAD /files/namespace/directory/file.ext
 
-Response:
+The response should be:
+
+```http
 Code: 200
 Content-Type: application/pdf
 Content-Length: 12345
-Libero-file-tags: "key=value,key2=value2,filename=bar.pdf"
-Link: "<http://sharedlink>;rel=????"
-Link: "<http://publiclink>;rel=????"
+Libero-file-tags: "filename=original-file.pdf,key1=value=1"
+Link: "<http://storage-url/path/to/file>;rel=https://libero.pub/public"
 Last-Modified: 2019-08-19T14:40:04.123456
 ETag: <etag>
+Vary: Prefer
 ```
 
-## Get file directly
-> What is this?
+The Link value is a direct url to the stored file.
 
+### Private storage
+
+If the underlying storage is private, then the following request should be used
+
+```http
+HEAD /files/libero/directory/file.pdf
+Prefer: link=signed
```

Comment:
shouldn't this be `link=https://libero.pub/signed`?

Reply:
Could be for consistency (might have to be quoted?), but nothing more.